src/mapping/impl/RBFParameterTuner.hpp

```diff
@@ -0,0 +1,196 @@
+#pragma once
+
+#include <mapping/MathHelper.hpp>
+#include <mapping/impl/BasisFunctions.hpp>
+#include <mesh/Mesh.hpp>
+
+namespace precice::mapping {
+
+struct RBFErrorEstimate {
+  double error;
+  double rcond;
+
+  RBFErrorEstimate(double error, double rcond)
+      : error(error), rcond(rcond)
+  {
+  }
+};
+
+struct Sample {
+  double radius;
+  double error;
+
+  Sample(double radius, double error)
+      : radius(radius), error(error) {}
+
+  Sample()
+      : radius(std::numeric_limits<double>::quiet_NaN()), error(std::numeric_limits<double>::max())
+  {
+  }
+};
+
+class RBFParameterTuner {
+
+  mutable logging::Logger _log{"mapping::RBFParameterTuner"};
+
+  /// Factor by which the radius is allowed to change before stopping
+  static constexpr double POS_TOLERANCE = 1.5;
+  /// Factor by which the error is allowed to change before stopping
+  static constexpr double ERR_TOLERANCE = 0.5;
+  /// Which reciprocal condition number should be considered too low for a good support radius
+  static constexpr double RCOND_TOLERANCE = 1e-12;
+  /// Number of iterations during the "bisection" step. After finding initial samples
+  static constexpr int MAX_BISEC_ITERATIONS = 6;
+
+protected:
+  std::vector<Sample> _samples;
+
+  double _lowerBound;
+  double _upperBound;
+  bool   _lastSampleWasOptimum;
+
+  Sample _currentOptimum;
+
+  static double estimateMeshResolution(mesh::Mesh &inputMesh);
+
+public:
+  ~RBFParameterTuner() = default;
+  explicit RBFParameterTuner(mesh::Mesh &inputMesh);
+
+  template <typename IndexContainer, typename Solver>
+  Sample optimize(Solver &solver, const IndexContainer &inputIds, const Eigen::VectorXd &inputData);
+
+  bool lastSampleWasOptimum() const;
+
+  double getLastOptimizedRadius() const;
+  double getLastOptimizationError() const;
+
+private:
+  static bool shouldContinue(const Sample &lowerBound, const Sample &upperBound);
+};
+
+inline RBFParameterTuner::RBFParameterTuner(mesh::Mesh &inputMesh)
+{
+  _lowerBound = estimateMeshResolution(inputMesh);
+  _upperBound = std::numeric_limits<double>::quiet_NaN();
+
+  _lastSampleWasOptimum = false;
+}
+
+template <typename IndexContainer, typename Solver>
+Sample RBFParameterTuner::optimize(Solver &solver, const IndexContainer &inputIds, const Eigen::VectorXd &inputData)
+{
+  constexpr bool radiusRBF = Solver::BASIS_FUNCTION_T::hasCompactSupport();
+  static_assert(radiusRBF, "RBF is not supported by this optimizer, as it does not accept a support-radius."
+                           "Currently supported: Compactly supported RBFs and Gaussians.");
+
+  // TODO: maybe make dependent on domain extend
+  constexpr double increaseSize = 10;
+  double           sampleRadius = _lowerBound;
+
+  _lastSampleWasOptimum = false;
+
+  PRECICE_INFO("Starting optimization with lower bound = {:.4e}, upper bound = {:.4e}", _lowerBound, _upperBound);
```

Comment:
This and other INFO prints will be confusing, if multiple mappings and participants are configured: to which mapping does it then refer?